tls/s2n_psk.h

```diff
@@ -89,9 +69,24 @@ int s2n_psk_calculate_binder(struct s2n_psk *psk, const struct s2n_blob *binder_
 int s2n_psk_verify_binder(struct s2n_connection *conn, struct s2n_psk *psk,
         const struct s2n_blob *partial_client_hello, struct s2n_blob *binder_to_verify);
 
-typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, 
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn,
                                           struct s2n_psk_identity *identities, size_t identities_length,
-                                          uint16_t *chosen_wire_index);                     
+                                          uint16_t *chosen_wire_index);
 /* This function will be labeled S2N_API and become a publicly visible api once we release the psk API. */
 int s2n_config_set_psk_selection_callback(struct s2n_connection *conn, s2n_psk_selection_callback cb);
 
+/* Public Interface -- will be made visible and moved to s2n.h when the PSK feature is released */
+
+struct s2n_psk;
+typedef enum {
+    S2N_PSK_HMAC_SHA224 = 0,
+    S2N_PSK_HMAC_SHA256,
+    S2N_PSK_HMAC_SHA384,
+} s2n_psk_hmac;
+
+struct s2n_psk* s2n_external_psk_new();
+int s2n_psk_free(struct s2n_psk **psk);
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_psk **psk_list, uint32_t psk_list_len);
```

Comment:
Since we're requiring that the caller pass in a list of PSKs, would it be worth doing a single `s2n_psk_list`? The API might look something like:

```c
struct s2n_psk_list* s2n_external_psk_new_list(uint16_t count);

int s2n_psk_list_set_identity(struct s2n_psk_list *psk, uint16_t index, const uint8_t *identity, uint16_t identity_size);
int s2n_psk_list_set_secret(struct s2n_psk_list *psk, uint16_t index, const uint8_t *secret, uint16_t secret_size);
int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_psk_list *psk_list);
```

This way we're only doing a single allocation for all PSKs.

Reply:
At that point, it might make more sense to just give up on the concept of "adding a list" at all and just have APIs to add (and maybe remove?) PSKs from the connection's list.

Reply:
New revision has given up on a list API :)